### PR TITLE
mDNS fix and env changes

### PR DESCRIPTION
**Pull request by jeremypoulter**
ArduinoOTA calls MDNS.begin with no hostname so removed the configured name. Now sets the same hostname as the main mDNS config.

Also added some changes the Platform IO envs;
- Added env to that will upload OTA from the IDE. 
- Removed some of the duplication in the different configs using the template feature of PlatformIO 3.1.0
- Use library names for the dependency

### CODE REVIEW

**Comment by jeremypoulter** [comment]
Now also includes a fix for the deploy end that was broken by Platform IO 3.1.0

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,5 @@
 .pioenvs
 .clang_complete
 .gcc-flags.json
-*~
+.piolibdeps
+*~
```
Comment (glynhudson) [inline .gitignore]: What's the problem with having git ignore gedit's tempory files?

File: platformio.ini [hunk 2]
```diff
@@ -21,35 +21,44 @@
...
 data_dir = src/data
 
+[common]
+version = -DBUILD_TAG=2.0.0
+#MQTT: https://github.com/knolleary/pubsubclient
+lib_deps = PubSubClient@2.6
+
 [env:emonesp]
-platform = espressif
+platform = espressif8266
```
Comment (glynhudson) [inline platformio.ini]: is the change `espressif > espressif8266` due to a platformIO change? Is there any difference? I can only see espressif8266 listed on the platformio platform page so I'm assuming it's just been renamed? http://platformio.org/platforms/espressif8266. But `espressif` still seems to work, it's probably a good idea to update my other projects that use this platform.

File: platformio.ini [hunk 3]
```diff
@@ -21,35 +21,44 @@
 env_default = emonesp
 data_dir = src/data
 
+[common]
+version = -DBUILD_TAG=2.0.0
+#MQTT: https://github.com/knolleary/pubsubclient
+lib_deps = PubSubClient@2.6
```
Comment (glynhudson) [inline platformio.ini]: Awesome! 👍  Love the new PlatformIO library manager. It's a step change in firmware build future replicability to be able to directly reference library versions. 

As I side note I have been in discussion with Ivan from platformIO regarding the pros and cons of using library ID vs Names vs direct GitHub reference: https://community.platformio.org/t/3-1-0-library-manager-installing-from-git/806

File: src/ota.cpp [hunk 4]
```diff
@@ -46,9 +46,12 @@ const char* u_host = "217.9.195.227";
 const char* u_url = "/esp/firmware.php";
 const char* firmware_update_path = "/upload";
 
+extern const char *esp_hostname;
+
 void ota_setup()
 {
   // Start local OTA update server
+  ArduinoOTA.setHostname(esp_hostname);
```
Comment (glynhudson) [inline src/ota.cpp]: What difference will this make? I haven't been using Arduino OTA therefore I'm probably unaware of any bugs here. Does this change fix or improve Arduino OTA? Have you been using Arduino OTA? Wondering if we should disable ArduinoOTA for production builds? I.e once we start using EmonESP for general use on OEM units.

File: platformio.ini (hunk 3 above)
Comment (jeremypoulter) [inline platformio.ini]: Personally I have a pet hate of magic numbers so I always try to remove them ;-)

File: platformio.ini (hunk 2 above)
Comment (jeremypoulter) [inline platformio.ini]: The ESP32 has now been released (very cool chip BTW  so the espressif platform was changed to espressif8266 (you get a warning when building with espressif)

File: .gitignore (hunk 1 above)
Comment (jeremypoulter) [inline .gitignore]: Nothing, just the was the diff is displayed, maybe I added a newline at the end, the line is still there.

File: src/ota.cpp (hunk 4 above)
Comment (jeremypoulter) [inline src/ota.cpp]: This probably is a result of the re-order I did, but ArduinoOTA calls MDNS.begin and resets the hostname, this change just ensures that no matter what order they are called in the same host name is set.

**Comment by jeremypoulter** [comment]
I solved the same problem in a different way;

```
src_build_flags = !(test -z $TRAVIS_TAG && echo '${common.version}') || echo '-DBUILD_TAG='$TRAVIS_TAG
```

This is essentially the same as you had previously but expanded out, rather than using the Bash short hand.

**Comment by glynhudson** [comment]
Great! Thanks a lot for answering my comments. Happy with all this :+1:

**Pull request merged**